src/helm/proxy/clients/palmyra_client.py

```diff
@@ -0,0 +1,184 @@
+from dataclasses import asdict
+
+
+class PalmyraClient(Client):
+    VALID_MODELS: List[str] = ["palmyra-base", "palmyra_large", "palmyra-r"]
+    TOKENIZER_NAME_PREFIX = "Writer/"
+
+    @staticmethod
+    def _get_model_name(model_name: str) -> str:
+        return "/".join(model_name.split("/")[1:])
+
+    def __init__(self, api_key: str, cache_config: CacheConfig):
+        self.api_key: str = api_key
+        self.cache = Cache(cache_config)
+        self._tokenizer_name_to_tokenizer: Dict[str, Tokenizer] = {}
+
+    def _get_tokenizer(self, tokenizer_name: str) -> Tokenizer:
+        modified_name = tokenizer_name.replace(self.TOKENIZER_NAME_PREFIX, "")
+        if modified_name not in self.VALID_MODELS:
+            raise ValueError(f"Invalid tokenizer: {tokenizer_name}")
+
+        # Check if the tokenizer is cached
+        if modified_name not in self._tokenizer_name_to_tokenizer:
+            self._tokenizer_name_to_tokenizer[modified_name] = AutoTokenizer.from_pretrained(
+                self.TOKENIZER_NAME_PREFIX + modified_name, use_fast=False
+            )
+            hlog(f"Initialized tokenizer: {modified_name}")
+        return self._tokenizer_name_to_tokenizer[modified_name]
+
+    def _send_request(self, model_name: str, raw_request: Dict[str, Any]) -> Dict[str, Any]:
+        response = requests.request(
+            method="POST",
+            url=f"https://enterprise-api.writer.com/llm/organization/3002/model/{model_name}/completions",
+            headers={
+                "Content-Type": "application/json",
+                "Accept": "application/json",
+                "Authorization": f"Bearer {self.api_key}",
+            },
+            data=json.dumps(raw_request),
+        )
+        result = json.loads(response.text)
+        assert "error" not in result, f"Request failed with error: {result['error']}"
+        return result
+
+    def make_request(self, request: Request) -> RequestResult:
+        """Make a request"""
+        raw_request = {
+            "prompt": request.prompt,
+            "minTokens": "1",  # Setting to 1 for now
+            "maxTokens": request.max_tokens,
+            "temperature": request.temperature,
+            "topP": request.top_p,
+            "topK": request.top_k_per_token,
+            "stop": request.stop_sequences,
+            "presencePenalty": request.presence_penalty,
+            "frequencyPenalty": request.frequency_penalty,
+            # bestOf: figure out what this does
+        }
+
+        completions: List[Sequence] = []
+        model_name: str = request.model_engine
+
+        # `num_completions` is not supported, so instead make `num_completions` separate requests.
+        for completion_index in range(request.num_completions):
+            try:
+
+                def do_it():
+                    result = self._send_request(model_name, raw_request)
+                    assert "choices" in result, f"Invalid response: {result}"
```

Comment:
Likewise, raise an exception instead of asserting.

Reply:
Done